### PR TITLE
Fix ART Blit detection regression and add D24S8 <-> RGBA8 conv to Vulkan

### PR DESCRIPTION
**Pull request by FernandoS27**
This fixes many issues in Pokemon Brilliant Diamond. Including, Vulkan & OGL crashes.

It's not in a good state to merge though.

### CODE REVIEW

File: src/video_core/host_shaders/convert_abgr8_to_d24s8.frag [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+layout(binding = 0) uniform sampler2D color_texture;
+
+void main() {
+    ivec2 coord = ivec2(gl_FragCoord.xy);
+    uvec4 color = uvec4(texelFetch(color_texture, coord, 0).rgba * (exp2(8) - 1.0f));
+    uint depth_unorm = (color.r << 16) | (color.g << 8) | color.b;
+
+    gl_FragDepth = float(depth_unorm) / (exp2(24.0) - 1.0f);
+    // gl_FragStencilRefARB = int(color.a);
```
Comment (FernandoS27) [inline src/video_core/host_shaders/convert_abgr8_to_d24s8.frag:16]: Note this shall be commented once a bufferCopy method has been implemented for NVIDIA.

**Comment by vonchenplus** [comment]
lol, It doesn't look like a simple binary copy.

**Comment by FernandoS27** [comment]
> lol, It doesn't look like a simple binary copy.

oh yeah, I forgot to mention that the technique comes from Citra.

File: src/video_core/renderer_vulkan/blit_image.cpp [hunk 2]
```diff
@@ -636,4 +749,62 @@ void BlitImageHelper::ConvertColorToDepthPipeline(vk::Pipeline& pipeline, VkRend
...
+        .flags = 0,
+        .stageCount = static_cast<u32>(stages.size()),
+        .pStages = stages.data(),
+        .pVertexInputState = &PIPELINE_VERTEX_INPUT_STATE_CREATE_INFO,
+        .pInputAssemblyState = &PIPELINE_INPUT_ASSEMBLY_STATE_CREATE_INFO,
+        .pTessellationState = nullptr,
+        .pViewportState = &PIPELINE_VIEWPORT_STATE_CREATE_INFO,
+        .pRasterizationState = &PIPELINE_RASTERIZATION_STATE_CREATE_INFO,
+        .pMultisampleState = &PIPELINE_MULTISAMPLE_STATE_CREATE_INFO,
+        .pDepthStencilState = &PIPELINE_DEPTH_STENCIL_STATE_CREATE_INFO,
```
Comment (ameerj) [inline src/video_core/renderer_vulkan/blit_image.cpp:799]: Looks like this is the only difference between `ConvertPipelineColorTargetEx` and `ConvertPipelineDepthTargetEx`
These can be refactored to a single `ConvertPipelineEx` which takes a `bool is_target_depth` that populates this field.

File: src/video_core/renderer_vulkan/blit_image.cpp [hunk 3]
```diff
@@ -448,6 +460,59 @@ void BlitImageHelper::ConvertR16ToD16(const Framebuffer* dst_framebuffer,
     Convert(*convert_r16_to_d16_pipeline, dst_framebuffer, src_image_view, up_scale, down_shift);
 }
 
+void BlitImageHelper::ConvertABGR8ToD24S8(const Framebuffer* dst_framebuffer,
+                                          const ImageView& src_image_view, u32 up_scale,
+                                          u32 down_shift) {
+    ConvertPipelineDepthTargetEx(convert_abgr8_to_d24s8_pipeline, dst_framebuffer->RenderPass(),
+                                 convert_abgr8_to_d24s8_frag, true);
```
Comment (ameerj) [inline src/video_core/renderer_vulkan/blit_image.cpp:467]: Looks like all usages of `ConvertPipelineDepthTargetEx` pass in `true` as the last param. Is that needed?

File: src/video_core/renderer_vulkan/blit_image.cpp [hunk 4]
```diff
@@ -448,6 +460,59 @@ void BlitImageHelper::ConvertR16ToD16(const Framebuffer* dst_framebuffer,
...
+    ConvertPipelineDepthTargetEx(convert_r16g16_to_d24s8_pipeline, dst_framebuffer->RenderPass(),
+                                 convert_r16g16_to_d24s8_frag, true);
+    Convert(*convert_r16g16_to_d24s8_pipeline, dst_framebuffer, src_image_view, up_scale,
+            down_shift);
+}
+
+void BlitImageHelper::ConvertD24S8ToABGR8(const Framebuffer* dst_framebuffer,
+                                          ImageView& src_image_view, u32 up_scale, u32 down_shift) {
+    ConvertPipelineColorTargetEx(convert_d24s8_to_abgr8_pipeline, dst_framebuffer->RenderPass(),
+                                 convert_d24s8_to_abgr8_frag, false);
```
Comment (ameerj) [inline src/video_core/renderer_vulkan/blit_image.cpp:493]: Looks like all usages of `ConvertPipelineColorTargetEx` pass in false as the last param

File: src/video_core/renderer_vulkan/vk_texture_cache.cpp [hunk 5]
```diff
@@ -754,6 +768,163 @@ StagingBufferRef TextureCacheRuntime::DownloadStagingBuffer(size_t size) {
...
+    return false;
+}
+
+VkBuffer TextureCacheRuntime::GetTemporaryBuffer(size_t needed_size) {
+    const auto level = (8 * sizeof(size_t)) - std::countl_zero(needed_size - 1ULL);
+    if (buffer_commits[level]) {
+        return *buffers[level];
+    }
+    const auto new_size = Common::NextPow2(needed_size);
+    VkBufferUsageFlags flags = VK_BUFFER_USAGE_TRANSFER_SRC_BIT | VK_BUFFER_USAGE_TRANSFER_DST_BIT |
```
Comment (ameerj) [inline src/video_core/renderer_vulkan/vk_texture_cache.cpp:785]: static constexpr

**Review by bunnei — APPROVED** [review]

**Comment by vonchenplus** [comment]
It is true that there is more duplicate code.

**Pull request merged**